### ISSUE TITLE
Handle NaN and +/-Infinity properly in JSON

### ISSUE THREAD

**Issue by danielballan** [bluesky/tiled#95]
The JSON format has no way of representing NaN or +/- infinity. Python has a [strange default behavior](https://docs.python.org/3/library/json.html#infinite-and-nan-number-values) of writing _invalid_ JSON when it encounters these values, placing `NaN` or `Infinity` unquoted (!) in the output. Most (all?) JSON parsers fail when they encounter this because it is invalid JSON.

I think our best option here is to place _quoted_ `"Infinity"`, `"-Infinity"` or `"NaN"` in the JSON. JSON parsers will happily parse that. It will still be incumbent on the recipient to deal with the fact that an array may have a mixture of numbers like `1` and strings like `"Infinity"`. It's the best we can do. We do something similar with date objects: JSON has no support for a datetime type, so we convert to ISO 8601 strings for JSON.

1. Copy [this method from the default built-in `JSONEncoder`](https://github.com/python/cpython/blob/b108db63e02797a795840152b82cab9792fd3ed2/Lib/json/encoder.py#L204-L243) onto [our ` _NumpySafeJSONEncoder`](https://github.com/bluesky/tiled/blob/25eaeee3c1abf945d1fd4037caeccc7cb4b12fb0/tiled/server/core.py#L540) to overrride the default. Edit it to change `'Infinity'` to `'"Infinity"'` and so on. (Overriding this entire method is a fairly extreme and potentially brittle way of overriding the default behavior, but Python does not provide any other hook for doing so. In the future, we might look to third-party JSON encoders instead of using the Python built-in one, but that is a large space that would take some time to navigate, so I'd like to defer that for now.)
2. Move  `_NumpySafeJSONEncoder` to `tiled.utils.NumpySafeJSONEncoder` so that it is "public" and usable in modules outside the server.
3. Everywhere in tiled where we use `json.dumps(...)`, such as [here](https://github.com/bluesky/tiled/blob/main/tiled/structures/array.py#L167) and several other places, add the parameter `json.dumps(..., cls=NumpySafeJSONEncoder)`.

**Comment by danielballan** [comment]
Above, I said that we had to go to extreme lengths to patch the JSONEncoder. @jmaruland has discovered that it's even worse than I realized: we'd need to copy a large and complex internal function from Python's JSON encoding module. We'd end up with a copy of nearly the entire module, and it's highly complex code.

I _also_ said that maybe we should consider using a third-party JSON encoder instead of the built-in one. I suggested that this could wait until later. But now, knowing how involved the patch would actually be, I wonder if we'd do better to switch to a third party solution instead. It looks like the best option is [orjson](https://github.com/ijl/orjson) which is:

1. Faster. It depends on the task, but for an array of numpy float64, it is 16X faster than the built-in JSON encoder.
2. Correct. By default, it [encodes `inf`, `-inf` and `nan` as the JSON `null` type](https://github.com/ijl/orjson#float). I think this is actually better than our original idea of encoding them as strings. Information is lost, but JSON is a lossy format. If you want floating point numbers with `nan` in them, choose another format like msgpack or Arrow that supports those types. It encodes [datetimes as strings](https://github.com/ijl/orjson#float), same as we have done in Tiled, and it [supports numpy arrays](https://github.com/ijl/orjson#numpy) too.

In summary, it seems like it does everything we want out of the box, and it's faster as well. That justifies adding a dependency to `requirements-server.txt`, in my opinion.

To borrow an idea from our friend James Powell: the built-in `json` module is good for getting a project off the ground. It basically does the job, and it's conveniently built in. But once a project grows and starts optimizing performance --- or indeed _correctness_ when it comes to `+/-inf`, `nan` --- it's time to upgrade to a third-party solution.

**Comment by danielballan** [comment]
Closed by #110
